### DISCUSSION TITLE
How to speed-up feature extraction?

### DISCUSSION THREAD

**Discussion by desh2608** [lhotse-speech/lhotse#440] [Q&A] [answered]
Default feature extraction in Lhotse is taking a long time. How can I speed-up this process or perform on-the-fly feature extraction?

cc:@aarora8

**Comment by desh2608** [comment, marked as answer]
Answer by @pzelasko:

1. (simplest) increase the number of jobs for feature extraction
2. Use GPU to accelerate feature extraction; you will need to `pip install kaldifeat` for this though. See here for the relevant method + example in Lhotse: https://github.com/lhotse-speech/lhotse/blob/master/lhotse/cut.py#L3358
3. You can use on-the-fly feature extraction. It might require a bit more IO but possibly increasing num_workers in DataLoader will offset that, if you notice a slow-down in training. All the changes you need to do are here:
https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/tdnn_lstm_ctc/asr_datamodule.py#L259:L267

**Reply by pzelasko** [reply]
Another viable option:

 4. Return batches of waveforms from your dataset, and use [`lhotse.features.kaldi.layers.Wav2LogFilterBank`](https://github.com/lhotse-speech/lhotse/blob/6976baf0acbda44acdc5b5fe02e6b845a7449f0a/lhotse/features/kaldi/layers.py#L363) as a layer in your neural network. Here is an example how you'd do it with `K2SpeechRecognitionDataset`:

```python
from lhotse.dataset import AudioSamples, K2SpeechRecognitionDataset
dset = K2SpeechRecognitionDataset(
                ...,
                input_strategy=AudioSamples(),
            )
```

**Comment by csukuangfj** [comment]
Also cc @EmreOzkose (from https://github.com/k2-fsa/icefall/issues/101)
